File: sources/xcode/JLKernel/JLKernelTests/JLKernelTests.swift
//
//  JLKernelTests/JLKernelTests.swift
//
//  Created by [Camilo Castro (@clsource)](https://ninjas.cl) on 2026-09-06
//  Made with love in Chile.
//
//  Copyright (c) Jasonelle.com
//
//  This file is part of Jasonelle Project <https://jasonelle.com>.
//  Jasonelle Project is dual licensed. You can choose between AGPLv3 or MPLv2.
//  MPLv2 is only valid if the software has a unique Jasonelle Key which was purchased in official channels at https://jasonelle.com.
//
//  == AGPLv3
//  Jasonelle is free software: you can redistribute it and/or modify it under the terms of the Affero GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
//  Jasonelle is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the Affero GNU General Public License for more details.
//  You should have received a copy of the Affero GNU General Public License along with Jasonelle. If not, see <https://www.gnu.org/licenses/agpl-3.0.txt>.
//
//  == MPLv2 (Only valid if purchased a Jasonelle Key)
//  This Source Code Form is subject to the terms
//  of the Mozilla Public License, v. 2.0.
//  If a copy of the MPL was not distributed
//  with this file, You can obtain one at
//
//  <https://mozilla.org/MPL/2.0/>.

import Foundation
import Testing
import WebKit
@testable import JLKernel

@Suite(.serialized) struct JLKernelTests {

    final class StubPlugin: JLKernel.Plugin {
        var receivedArgs: Any?
        var receivedCallbackId: String?

        override public func handle_call(callbackId: String, args: [String: Any]? = [:], respond: @escaping (String) -> Void) {
            self.receivedArgs = args
            self.receivedCallbackId = callbackId
        }
    }

    @Test @MainActor func dispatchesToRegisteredPlugin() async throws {
        let plugin = StubPlugin()
        let webview = JLKernel.WebView(config: AppConfiguration(url: URL(string: "https://jasonelle.com")!), plugins: ["stub": plugin])
        let coordinator = webview.makeCoordinator()

        coordinator.handleMessage(body: ["name": "stub", "args": ["value": 42], "callbackId": "call_1"])

        #expect((plugin.receivedArgs as? [String: Any])?["value"] as? Int == 42)
        #expect(plugin.receivedCallbackId == "call_1")
    }

    @Test @MainActor func ignoresUnknownPlugins() async throws {
        let plugin = StubPlugin()
        let webview = JLKernel.WebView(config: AppConfiguration(url: URL(string: "https://jasonelle.com")!), plugins: ["stub": plugin])
        let coordinator = webview.makeCoordinator()

        coordinator.handleMessage(body: ["name": "missing", "args": [:]])
        coordinator.handleMessage(body: "not a dictionary")

        #expect(plugin.receivedArgs == nil)
    }

}

// MARK: - Ratlog

struct RatlogTests {

    @Test func formatsTagsMessageAndSortedFields() {
        let output = Ratlog.format(message: "hello", tags: ["b", "a"], fields: ["z": "1", "a": "2"])

        #expect(output == "[b|a] hello | a: 2 | z: 1")
    }

    @Test func omitsEmptyTagsAndFields() {
        #expect(Ratlog.format(message: "hi", tags: [], fields: [:]) == "hi")
    }

    @Test func omitsTagsOnly() {
        #expect(Ratlog.format(message: "hi", tags: [], fields: ["a": "1"]) == "hi | a: 1")
        #expect(Ratlog.format(message: "hi", tags: ["t"], fields: [:]) == "[t] hi")
    }

}

// MARK: - LogLevel

struct LogLevelTests {

    @Test func comparesBySeverity() {
        #expect(LogLevel.debug < LogLevel.info)
        #expect(LogLevel.info < LogLevel.notice)
        #expect(LogLevel.emergency > LogLevel.alert)
    }

}

// MARK: - License

struct LicenseTests {

    private func makeLicense(
        key: String? = "",
        isSimulator: Bool
    ) -> License {
        License(key: key) { isSimulator }
    }

    @Test func checkWithValidKeyDoesNotCrash() {
        makeLicense(key: "test-key", isSimulator: false).check()
    }

    @Test func checkWithEmptyKeyOnSimulatorDoesNotCrash() {
        makeLicense(isSimulator: true).check()
    }

    @Test func abortReturnsOnSimulator() {
        makeLicense(isSimulator: true).abortIfIsInSimulator()
    }

    @Test func checkWithNilKeyOnSimulatorDoesNotCrash() {
        makeLicense(key: nil, isSimulator: true).check()
    }

    @Test func checkWithPurchaseMeKeyOnSimulatorDoesNotCrash() {
        makeLicense(key: "PURCHASE_ME", isSimulator: true).check()
    }

    @Test func checkWithValidKeyOnSimulatorDoesNotCrash() {
        makeLicense(key: "real-key", isSimulator: true).check()
    }

    @Test func verifyWithKeyDoesNotCrash() {
        License.verify(key: "test-key")
    }

    @Test func verifyWithEmptyKeyOnSimulatorDoesNotCrash() {
        License.verify(key: "")
    }

}

// MARK: - Version

struct VersionTests {

    @Test func semanticReturnsTrimmedBundledVersion() {
        let version = Version.semantic()

        #expect(!version.isEmpty)
        #expect(version == version.trimmingCharacters(in: .whitespacesAndNewlines))
    }

    @Test func semanticHonorsOverride() {
        Version.version = "9.9.9-test"
        defer { Version.version = nil }

        #expect(Version.semantic() == "9.9.9-test")
    }

}

// MARK: - Event

final class EventStubPlugin: JLKernel.Plugin {
    var receivedNames: [String] = []

    override func handle_event(_ event: String, args: [String: Any]? = [:], respond: @escaping (String) -> Void) {
        receivedNames.append(event)
        respond("")
    }
}

struct EventTests {

    @Test func onAppearSendDispatchesToRegisteredPlugins() {
        let plugins: [String: JLKernel.Plugin] = [
            "a": EventStubPlugin(),
            "b": EventStubPlugin()
        ]
        Events.plugins = plugins
        defer { Events.plugins = [:] }

        JLKernel.Events.sendOnAppear()

        let names = plugins.values.compactMap { ($0 as? EventStubPlugin)?.receivedNames }.flatMap { $0 }
        #expect(names.count == 2)
        #expect(names.allSatisfy { $0 == Events.contentViewOnAppear.rawValue })
    }

    @Test func onAppearSendWithNoPluginsIsNoop() {
        Events.plugins = [:]
        defer { Events.plugins = [:] }

        JLKernel.Events.sendOnAppear()
    }

}

// MARK: - Plugin

final class UnconfiguredPlugin: JLKernel.Plugin {}

struct PluginTests {

    @Test func defaultNameIsTypeName() {
        #expect(UnconfiguredPlugin.name == "UnconfiguredPlugin")
    }

    @Test func defaultCallDoesNotRespond() {
        var response: String?

        UnconfiguredPlugin().handle_call(callbackId: "call_1", args: nil) { response = $0 }

        #expect(response == nil)
    }

    @Test func defaultEventDoesNotRespond() {
        var response: String?

        UnconfiguredPlugin().handle_event("viewDidLoad", args: nil) { response = $0 }

        #expect(response == nil)
    }

    @Test func jsLoadsBundledPluginJS() {
        let source = UnconfiguredPlugin().js()

        #expect(source.contains("window.jasonelle.plugins.stub"))
    }

}

// MARK: - Navigation policy (allowed hosts)

struct NavigationPolicyTests {

    private let mainURL = URL(string: "https://jasonelle.com")!

    @MainActor private func makeCoordinator() -> Coordinator {
        JLKernel.WebView(config: AppConfiguration(url: mainURL)).makeCoordinator()
    }

    @Test @MainActor func allowsEveryURLWhenAllowedIsEmptyOrNil() async throws {
        let coordinator = makeCoordinator()

        #expect(coordinator.decidePolicy(url: URL(string: "https://anything.com"), allowed: [], mainURL: mainURL) == .allow)
        #expect(coordinator.decidePolicy(url: URL(string: "https://anything.com"), allowed: nil, mainURL: mainURL) == .allow)
        #expect(coordinator.decidePolicy(url: nil, allowed: nil, mainURL: mainURL) == .allow)
    }

    @Test @MainActor func allowsURLsWhoseHostIsInAllowedList() async throws {
        let coordinator = makeCoordinator()

        #expect(coordinator.decidePolicy(url: URL(string: "https://jasonelle.com/foo?bar=1"), allowed: ["jasonelle.com"], mainURL: mainURL) == .allow)
    }

    @Test @MainActor func cancelsURLsWhoseHostIsNotInAllowedList() async throws {
        let coordinator = makeCoordinator()

        #expect(coordinator.decidePolicy(url: URL(string: "https://evil.com"), allowed: ["jasonelle.com"], mainURL: mainURL) == .cancel)
        #expect(coordinator.decidePolicy(url: URL(string: "file:///tmp/x"), allowed: ["jasonelle.com"], mainURL: mainURL) == .cancel)
    }

    @Test @MainActor func alwaysAllowsTheAppURL() async throws {
        let coordinator = makeCoordinator()

        #expect(coordinator.decidePolicy(url: mainURL, allowed: ["other.com"], mainURL: mainURL) == .allow)
    }

}

// MARK: - WebView user script injection (plugins + webview.js)

struct WebViewScriptInjectionTests {

    @Test func bridgePostsReturnPromiseRoutedByCallbackId() {
        let source = JLKernel.WebView.jsBridgeScript

        #expect(source.contains("new Promise"))
        #expect(source.contains("callbackId"))
        #expect(source.contains("postMessage"))
    }

    @Test @MainActor func injectsWebViewJSAfterPluginScripts() async throws {
        let webview = JLKernel.WebView(
            config: AppConfiguration(url: URL(string: "https://jasonelle.com")!),
            plugins: ["stub": JLKernelTests.StubPlugin()]
        )
        let target = WKWebView()

        // The bundle that contains the test fixtures (this mock webview.js)
        let bundle = Bundle(for: JLKernelTests.StubPlugin.self)

        webview.injectUserScripts(into: target, bundle: bundle)

        let sources = target.configuration.userContentController.userScripts.map(\.source)

        // The plugin script from the test bundle is injected first
        #expect(sources.contains { $0.contains("window.jasonelle.plugins.stub") })
        // The mocked webview.js is injected last, after the plugin scripts
        #expect(sources.last?.contains("JLKernelTests webview.js") == true)
        #expect(sources.dropLast().contains { $0.contains("window.jasonelle.plugins.stub") })
    }

}

// MARK: - ConfigurationLoader

struct ConfigurationLoaderTests {

    @Test func decodeReturnsURLFromValidJSON() throws {
        let json = #"{"url": "https://example.com"}"#
        let data = Data(json.utf8)

        let config = try ConfigurationLoader.decode(data: data)

        #expect(config.url == URL(string: "https://example.com"))
    }

    @Test func decodeStripsJSONCComments() throws {
        let json = #"{/* comment */"url": "https://example.com"}"#
        let data = Data(json.utf8)

        let config = try ConfigurationLoader.decode(data: data)

        #expect(config.url == URL(string: "https://example.com"))
    }

    @Test func decodeThrowsOnInvalidJSON() {
        let data = Data("not json".utf8)

        #expect(throws: ConfigurationError.self) {
            try ConfigurationLoader.decode(data: data)
        }
    }

    @Test func loadThrowsFileNotFoundWhenURLIsNil() {
        #expect(throws: ConfigurationError.self) {
            try ConfigurationLoader.load(from: nil)
        }
    }

    @Test func loadReadsConfigFile() throws {
        let configURL = URL(fileURLWithPath: #file)
            .deletingLastPathComponent()
            .appendingPathComponent("config.jsonc")

        let config = try ConfigurationLoader.load(from: configURL)

        #expect(config.url == URL(string: "https://jasonelle.com"))
    }

}

File: sources/android/JLKernel/src/test/java/com/jasonelle/kernel/JLKernelTests.kt
//
//  JLKernelTests/JLKernelTests.kt
//
//  Created by [Camilo Castro (@clsource)](https://ninjas.cl) on 2026-09-15
//  Made with love in Chile.
//
//  Copyright (c) Jasonelle.com
//
//  This file is part of Jasonelle Project <https://jasonelle.com>.
//  Jasonelle Project is dual licensed. You can choose between AGPLv3 or MPLv2.
//  MPLv2 is only valid if the software has a unique Jasonelle Key which was purchased in official channels at https://jasonelle.com.
//
//  == AGPLv3
//  Jasonelle is free software: you can redistribute it and/or modify it under the terms of the Affero GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
//  Jasonelle is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the Affero GNU General Public License for more details.
//  You should have received a copy of the Affero GNU General Public License along with Jasonelle. If not, see <https://www.gnu.org/licenses/agpl-3.0.txt>.
//
//  == MPLv2 (Only valid if purchased a Jasonelle Key)
//  This Source Code Form is subject to the terms
//  of the Mozilla Public License, v. 2.0.
//  If a copy of the MPL was not distributed
//  with this file, You can obtain one at
//
//  <https://mozilla.org/MPL/2.0/>.

package com.jasonelle.kernel

import java.io.ByteArrayInputStream
import java.nio.charset.StandardCharsets
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Test

class JLKernelTests {

    class StubPlugin : Plugin() {
        var receivedArgs: Map<String, Any?>? = null
        var receivedCallbackId: String? = null

        override fun handle_call(callbackId: String, args: Map<String, Any?>?, respond: (String) -> Unit) {
            receivedArgs = args
            receivedCallbackId = callbackId
        }
    }

    @Test
    fun dispatchesToRegisteredPlugin() {
        val plugin = StubPlugin()
        val config = AppConfiguration(urlString = "https://jasonelle.com")
        val coordinator = Coordinator(config, mapOf("stub" to plugin))

        coordinator.handleMessage("stub", "call_1", mapOf("value" to 42))

        assertEquals(42, plugin.receivedArgs?.get("value"))
        assertEquals("call_1", plugin.receivedCallbackId)
    }

    @Test
    fun ignoresUnknownPlugins() {
        val plugin = StubPlugin()
        val config = AppConfiguration(urlString = "https://jasonelle.com")
        val coordinator = Coordinator(config, mapOf("stub" to plugin))

        coordinator.handleMessage("missing", "call_1", emptyMap())

        assertNull(plugin.receivedArgs)
        assertNull(plugin.receivedCallbackId)
    }
}

class RatlogTests {

    @Test
    fun formatsTagsMessageAndSortedFields() {
        val output = Ratlog.format("hello", listOf("b", "a"), mapOf("z" to "1", "a" to "2"))

        assertEquals("[b|a] hello | a: 2 | z: 1", output)
    }

    @Test
    fun omitsEmptyTagsAndFields() {
        assertEquals("hi", Ratlog.format("hi", emptyList(), emptyMap()))
    }

    @Test
    fun omitsTagsOnly() {
        assertEquals("hi | a: 1", Ratlog.format("hi", emptyList(), mapOf("a" to "1")))
        assertEquals("[t] hi", Ratlog.format("hi", listOf("t"), emptyMap()))
    }
}

class LogLevelTests {

    @Test
    fun comparesBySeverity() {
        assertTrue(LogLevel.DEBUG < LogLevel.INFO)
        assertTrue(LogLevel.INFO < LogLevel.NOTICE)
        assertTrue(LogLevel.EMERGENCY > LogLevel.ALERT)
    }
}

class LicenseTests {

    private fun makeLicense(key: String? = "", isSimulator: Boolean): License {
        return License(key = key) { isSimulator }
    }

    @Test
    fun checkWithValidKeyDoesNotCrash() {
        makeLicense(key = "test-key", isSimulator = false).check()
    }

    @Test
    fun checkWithEmptyKeyOnSimulatorDoesNotCrash() {
        makeLicense(isSimulator = true).check()
    }

    @Test
    fun abortReturnsOnSimulator() {
        makeLicense(isSimulator = true).abortIfIsInSimulator()
    }

    @Test
    fun checkWithNilKeyOnSimulatorDoesNotCrash() {
        makeLicense(key = null, isSimulator = true).check()
    }

    @Test
    fun checkWithPurchaseMeKeyOnSimulatorDoesNotCrash() {
        makeLicense(key = "PURCHASE_ME", isSimulator = true).check()
    }

    @Test
    fun checkWithValidKeyOnSimulatorDoesNotCrash() {
        makeLicense(key = "real-key", isSimulator = true).check()
    }

    @Test
    fun checkWithEmptyKeyOnDeviceThrows() {
        try {
            makeLicense(isSimulator = false).check()
            fail("Expected IllegalStateException")
        } catch (_: IllegalStateException) {
            // expected
        }
    }
}

class VersionTests {

    @Test
    fun semanticReturnsTrimmedBundledVersion() {
        val version = Version.semantic()

        assertTrue(version.isNotEmpty())
        assertEquals(version, version.trim())
    }

    @Test
    fun semanticHonorsOverride() {
        Version.version = "9.9.9-test"
        try {
            assertEquals("9.9.9-test", Version.semantic())
        } finally {
            Version.version = null
        }
    }
}

class EventStubPlugin : Plugin() {
    val receivedNames = mutableListOf<String>()

    override fun handle_event(event: String, args: Map<String, Any?>?, respond: (String) -> Unit) {
        receivedNames.add(event)
        respond("")
    }
}

class EventTests {

    @Test
    fun onAppearSendDispatchesToRegisteredPlugins() {
        val plugins: Map<String, Plugin> = mapOf(
            "a" to EventStubPlugin(),
            "b" to EventStubPlugin()
        )
        Events.register(plugins)
        try {
            Events.sendOnAppear()

            val names = plugins.values.map { (it as EventStubPlugin).receivedNames }.flatten()
            assertEquals(2, names.size)
            assertTrue(names.all { it == Events.CONTENT_VIEW_ON_APPEAR.rawValue })
        } finally {
            Events.register(emptyMap())
        }
    }

    @Test
    fun onAppearSendWithNoPluginsIsNoop() {
        Events.register(emptyMap())
        try {
            Events.sendOnAppear()
        } finally {
            Events.register(emptyMap())
        }
    }
}

class UnconfiguredPlugin : Plugin()

class PluginTests {

    @Test
    fun defaultNameIsTypeName() {
        assertEquals("UnconfiguredPlugin", UnconfiguredPlugin().name)
    }

    @Test
    fun defaultCallDoesNotRespond() {
        var response: String? = null

        UnconfiguredPlugin().handle_call("call_1", null) { response = it }

        assertNull(response)
    }

    @Test
    fun defaultEventDoesNotRespond() {
        var response: String? = null

        UnconfiguredPlugin().handle_event("viewDidLoad", null) { response = it }

        assertNull(response)
    }

    @Test
    fun resolveBuildsPromiseScript() {
        var response: String? = null

        UnconfiguredPlugin().resolve(mapOf("message" to "Hello"), "call_1", respond = { response = it })

        assertNotNull(response)
        val script = response ?: return
        assertTrue(script.startsWith("window.jasonelle.result.resolve({"))
        assertTrue(script.contains("\"status\":\"ok\""))
        assertTrue(script.contains("\"callbackId\":\"call_1\""))
        assertTrue(script.contains("\"message\":\"Hello\""))
    }

    @Test
    fun rejectBuildsPromiseScript() {
        var response: String? = null

        UnconfiguredPlugin().reject(mapOf("error" to "nope"), "call_1", status = "error", respond = { response = it })

        assertNotNull(response)
        val script = response ?: return
        assertTrue(script.startsWith("window.jasonelle.result.reject({"))
        assertTrue(script.contains("\"status\":\"error\""))
    }

    @Test
    fun jsLoadsBundledPluginJS() {
        val source = UnconfiguredPlugin().js()

        assertTrue(source.contains("window.jasonelle.plugins.stub"))
    }
}

class NavigationPolicyTests {

    private val mainHost = "jasonelle.com"

    @Test
    fun allowsEveryURLWhenAllowedIsEmptyOrNil() {
        val coordinator = Coordinator(AppConfiguration(urlString = "https://jasonelle.com"), emptyMap())

        assertEquals(NavigationPolicy.ALLOW, coordinator.decidePolicyForHost("anything.com", emptyList(), mainHost))
        assertEquals(NavigationPolicy.ALLOW, coordinator.decidePolicyForHost("anything.com", null, mainHost))
        assertEquals(NavigationPolicy.ALLOW, coordinator.decidePolicyForHost(null, null, mainHost))
    }

    @Test
    fun allowsURLsWhoseHostIsInAllowedList() {
        val coordinator = Coordinator(AppConfiguration(urlString = "https://jasonelle.com"), emptyMap())

        assertEquals(
            NavigationPolicy.ALLOW,
            coordinator.decidePolicyForHost("jasonelle.com", listOf("jasonelle.com"), mainHost)
        )
    }

    @Test
    fun cancelsURLsWhoseHostIsNotInAllowedList() {
        val coordinator = Coordinator(AppConfiguration(urlString = "https://jasonelle.com"), emptyMap())

        assertEquals(
            NavigationPolicy.CANCEL,
            coordinator.decidePolicyForHost("evil.com", listOf("jasonelle.com"), mainHost)
        )
        assertEquals(
            NavigationPolicy.CANCEL,
            coordinator.decidePolicyForHost(null, listOf("jasonelle.com"), mainHost)
        )
    }

    @Test
    fun alwaysAllowsTheAppURL() {
        val coordinator = Coordinator(AppConfiguration(urlString = "https://jasonelle.com"), emptyMap())

        assertEquals(
            NavigationPolicy.ALLOW,
            coordinator.decidePolicyForHost(mainHost, listOf("other.com"), mainHost)
        )
    }
}

class WebViewBridgeTests {

    @Test
    fun bridgePostsReturnPromiseRoutedByCallbackId() {
        val source = JasonelleBridge.JS_BRIDGE_SCRIPT

        assertTrue(source.contains("new Promise"))
        assertTrue(source.contains("callbackId"))
        assertTrue(source.contains("postMessage"))
        assertTrue(source.contains("window.jasonelle"))
    }
}

class ConfigurationLoaderTests {

    @Test
    fun decodeReturnsURLFromValidJSON() {
        val config = ConfigurationLoader.decode("""{"url": "https://example.com"}""")

        assertEquals("https://example.com", config.urlString)
    }

    @Test
    fun decodeStripsJSONCComments() {
        val config = ConfigurationLoader.decode(
            """
            {
              /* block comment */
              "url": "https://example.com", // line comment
              // another comment
              "inspectable": false,
              "allowed": ["example.com"]
            }
            """.trimIndent()
        )

        assertEquals("https://example.com", config.urlString)
        assertEquals(false, config.inspectable)
        assertEquals(listOf("example.com"), config.allowed)
    }

    @Test
    fun decodeThrowsOnInvalidJSON() {
        try {
            ConfigurationLoader.decode("not json")
            fail("Expected ConfigurationException")
        } catch (_: ConfigurationException) {
            // expected
        }
    }

    @Test
    fun stripJSONCCommentsKeepsCommentsInsideStrings() {
        val input = """{"url": "https://example.com/a//b", "note": "not /* a comment */"}"""
        val stripped = ConfigurationLoader.stripJSONCComments(input)

        assertTrue(stripped.contains("https://example.com/a//b"))
        assertTrue(stripped.contains("not /* a comment */"))
    }

    @Test
    fun loadThrowsFileNotFoundWhenStreamIsNull() {
        try {
            ConfigurationLoader.load(null)
            fail("Expected ConfigurationException")
        } catch (_: ConfigurationException) {
            // expected
        }
    }

    @Test
    fun loadReadsConfigStream() {
        val json = """{"url": "https://jasonelle.com"}"""
        val stream = ByteArrayInputStream(json.toByteArray(StandardCharsets.UTF_8))

        val config = ConfigurationLoader.load(stream)

        assertEquals("https://jasonelle.com", config.urlString)
    }
}